plugins/chat/app/services/chat/update_message.rb

```diff
@@ -0,0 +1,143 @@
+# frozen_string_literal: true
+
+module Chat
+  # Service responsible for updating a message.
+  #
+  # @example
+  #  Chat::UpdateMessage.call(message_id: 2, guardian: guardian, message: "A new message")
+  #
+  class UpdateMessage
+    include Service::Base
+
+    # @!method call(message_id:, guardian:, message:, upload_ids:)
+    #   @param guardian [Guardian]
+    #   @param message_id [Integer]
+    #   @param message [String]
+    #   @param upload_ids [Array<Integer>] IDs of uploaded documents
+
+    contract
+    model :message
+    model :uploads, optional: true
+    policy :can_modify_channel_message
+    policy :can_modify_message
+
+    transaction do
+      step :modify_message
+      step :save_message
+      step :save_revision
+      step :publish
+    end
+
+    class Contract
+      attribute :message_id, :string
+      attribute :message, :string
+      attribute :upload_ids, :array
+
+      validates :message_id, presence: true
+      validates :message, presence: true, if: -> { upload_ids.blank? }
+    end
+
+    private
+
+    def fetch_message(contract:, **)
+      ::Chat::Message
+        .strict_loading
+        .includes(
+          :chat_mentions,
+          :bookmarks,
+          :chat_webhook_event,
+          :uploads,
+          :revisions,
+          reactions: [:user],
+          thread: [:channel, last_message: [:user]],
+          chat_channel: [
+            :last_message,
+            :chat_channel_archive,
+            chatable: [:topic_only_relative_url, direct_message_users: [user: :user_option]],
+          ],
+          user: :user_status,
+        )
```

Comment:
I wonder if this is worth it, given it's only for one message, but I guess we will see...